### ISSUE TITLE
Unexpected increase of  imap::client::Connection.tag

### ISSUE THREAD

**Issue by qsdgy** [jonhoo/rust-imap#229]
imap::client::Connection.tag has been [increased ](https://github.com/jonhoo/rust-imap/blob/db2911746358741a74f9dda8b2eecbf480506224/src/client.rs#L1469)no matter whether writing command to socket is okay.
```rust
fn run_command(&mut self, untagged_command: &str) -> Result<()> {
        let command = self.create_command(untagged_command);
        self.write_line(command.into_bytes().as_slice())
}
```
```rust
fn create_command(&mut self, command: &str) -> String {
        self.tag += 1;
        format!("{}{} {}", TAG_PREFIX, self.tag, command)
}
```

If `self.write_line()` fails, the next imap command call would violate [assertion](https://github.com/jonhoo/rust-imap/blob/db2911746358741a74f9dda8b2eecbf480506224/src/client.rs#L1396), and program panics.

---
background:
Using TcpStream with read/write timeout configured as underlying stream, ~~read()~~ [__read/write__](https://man7.org/linux/man-pages/man7/socket.7.html#:~:text=if%20no%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20data%20has%20been%20transferred%20and%20the%20timeout%20has%20been%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20reached%2C%20then%20%2D1%20is%20returned%20with%20errno%20set%20to%20EAGAIN%20or%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20EWOULDBLOCK%2C%20or%20EINPROGRESS%20(for%20connect(2))%20just%20as%20if%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20the%20socket%20was%20specified%20to%20be%20nonblocking.) may retrun WouldBlock/TimedOut error.

If some command run to failed, I try to logout() and create new session. and panic in logout().
```
[2022-07-01T14:53:07Z ERROR my_crate] fetch_insert_store failed: [Other(Io(Os { code: 11, kind: WouldBlock, message: "Resource temporarily unavailable" }))]
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `[97, 51, 53]`,
 right: `[97, 51, 54]`', /path/to/imap-3.0.0-alpha.7/src/client.rs:1396:25
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by qsdgy** [comment]
one straight forward fix which won't touch tests:

```rust
fn run_command(&mut self, untagged_command: &str) -> Result<()> {
    let command = self.create_command(untagged_command);
    let result = self.write_line(command.into_bytes().as_slice());
    if result.is_err() {
        self.tag -= 1;
    }
    result
}
```

**Comment by jonhoo** [comment]
I'm not sure this is the behavior we want — in _general_ you'll want to increment the tag for the next request even if the previous request failed. For example, on a `NO` or `BAD`, I believe the server expectation will still be that you use a new tag for the next request. Unless the IMAP protocol permits re-using tags — I forget whether that is the case. I think we should _only_ reset the tag on a timeout, since any other operation should increment the tag.

Even timeout I'm unsure about. Timeouts are real weird, because they may have succeeded in writing _some_ of the data, but not other parts, which isn't something we're set up to deal with. I think the only safe way to deal with a timeout (at least the way the library is currently written) is to restart the session.

**Comment by qsdgy** [comment]
Surely not all write() errors can be treated as "client sent failed, server won't receive anything", I guess it's a common case. And I trust that when the ErrorKind is [WouldBlock](https://doc.rust-lang.org/stable/std/io/enum.ErrorKind.html#variant.WouldBlock), the client [doesn't perform any transfer](https://man7.org/linux/man-pages/man7/socket.7.html#:~:text=if%20no%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20data%20has%20been%20transferred%20and%20the%20timeout%20has%20been%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20reached%2C%20then%20%2D1%20is%20returned%20with%20errno%20set%20to%20EAGAIN%20or%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20EWOULDBLOCK%2C%20or%20EINPROGRESS%20(for%20connect(2))%20just%20as%20if%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20the%20socket%20was%20specified%20to%20be%20nonblocking.).

So there are two choices to deal with the write() failure:
* ignore it
If request indeed transfered to server even write() call return error, it's ok. otherwise corrupts tag.
* rollback tag
If request indeed transfered to server even write() call return error, it corrupts tag. otherwise it's ok.

I think the second one is more comfortable.

---
The request failed scenes maybe behind the successful write() branch.

**Comment by jonhoo** [comment]
Yeah, WouldBlock is, I think, possibly the _only_ case where rollback is okay. WouldBlock also shouldn't _generally_ happen in this crate since it is written assuming that the socket isn't non-blocking. I wonder if instead what we'll want is to keep the current behavior and then expose a function like `Client::rewind_tag` that a caller can invoke specifically if they _know_ that the last tag wasn't re-used. That could also then be used in cases where the caller knows that a timeout _didn't_ write anything and therefore that the tag is safe to re-use.

**Comment by qsdgy** [comment]
If someone (like poor me) doesn't lucky enough to operate with smoothly network, man has to using a tcpstream with [r/o timeout ](https://man7.org/linux/man-pages/man7/socket.7.html#:~:text=just%20as%20if%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20the%20socket%20was%20specified%20to%20be%20nonblocking)configured (which probably **should be the default option**, but that‘s another topic) to prevent socket from hanging forever. In this situation, EWOULDBLOCK is expected as normal.

SO_RCVTIMEO and SO_SNDTIMEO:
> if no data has been transferred and the timeout has been
              reached, then -1 is returned with errno set to EAGAIN or
              EWOULDBLOCK, or EINPROGRESS (for connect(2)) just as if
             **the socket was specified to be nonblocking**.

Above explains why `WouldBlock` happens even the socket isn't non-blocking.

Though `Client::rewind_tag` could be workaround, even without considing tcpstream with r/o timeout,  I wonder what's the decisive reason to assuming server successfully got the increased tag when write() is failed rather than opposite. In my opinion, even some function like `Client::forward_tag` invoked when send command fails and off-by-one tag detected is unnecessary.

**Comment by jonhoo** [comment]
Oh, that `SO_*TIMEO` man page quote is interesting, thank you!

The challenge is that a single top-level write to the socket isn't necessarily a single write _syscall_. For example, when using a `BufferedWriter` or going through a TLS writer, a high-level write of a string may very well end up being several writes (e.g., by line) to the underlying socket, where some prefix of them may succeed and the later fails. It seems safer to assume that the server _may_ have gotten the tag, even if an error was encountered, than to assume that it did not and continue sending bytes effectively under the same tag number. That said, using a client after an error is _always_ risky, so I'd be okay with only incrementing the tag if no errors occur and then providing a `Client::skip_tag` method for explicitly incrementing it :+1:

**Comment by qsdgy** [comment]
I'm not sure but seems that every `Connection::run_command` has corresponding `Connection::readline` follows, so maybe situation of `tag n` succeed but `tag n+1` fails will never happen? ( if `tag n` is bufferd, following `readline` cause deadlock ).

---
> explicitly incrementing it

I think that's just what we need.

**Comment by jonhoo** [comment]
Yep, incrementing on readline seems reasonable!

**Comment by jonhoo** [comment]
Fixed by #284 in alpha.13
